lib/ABI/FunctionType.cpp

```diff
@@ -52,74 +55,95 @@ bool verify(const SortedVector<RegisterType> &UsedRegisters,
 
-static std::optional<model::QualifiedType>
-buildDoubleType(model::Register::Values UpperRegister,
-                model::Register::Values LowerRegister,
-                model::PrimitiveTypeKind::Values CustomKind,
-                model::Binary &TheBinary) {
-  model::PrimitiveTypeKind::Values UpperKind = selectTypeKind(UpperRegister);
-  model::PrimitiveTypeKind::Values LowerKind = selectTypeKind(LowerRegister);
-  if (UpperKind != LowerKind)
-    return std::nullopt;
+  // Skip ABIs that do not allow returning big values.
+  // They do not benefit from this check.
+  if constexpr (LocationRegister != model::Register::Invalid) {
+    if constexpr (isVectorRegisterHelper(LocationRegister)) {
+      // Vector register used as the return value locations are not supported.
+      return false;
+    } else if constexpr (constexprIsContained(AT::CalleeSavedRegisters,
+                                              LocationRegister)) {
+      // Using callee saved register as a return value location doesn't make
+      // much sense: filter those out.
+        return false;
+    } else {
+      // The return value location register can optionally also be the first GPRs,
+      // but only the first one.
+      constexpr auto &GPRs = AT::GeneralPurposeArgumentRegisters;
+      constexpr auto Iterator = constexprFind(GPRs, LocationRegister);
+      if constexpr (Iterator != GPRs.end() && Iterator != GPRs.begin())
+        return false;
+    }
+  }
 
-  size_t UpperSize = model::Register::getSize(UpperRegister);
-  size_t LowerSize = model::Register::getSize(LowerRegister);
-  return model::QualifiedType(TheBinary.getPrimitiveType(CustomKind,
-                                                         UpperSize + LowerSize),
-                              {});
+  return true;
 }
 
-static model::QualifiedType getTypeOrDefault(const model::QualifiedType &Type,
-                                             model::Register::Values Register,
-                                             model::Binary &Binary) {
-  if (Type.UnqualifiedType.get() != nullptr)
-    return Type;
-  else
-    return buildType(Register, Binary);
-}
+/// \brief Replace all the references to `OldKey` with the references to
+///        the newly added `NewType`. Also erases the original type.
+///
+/// \param OldKey The type references of which should be replaced.
+/// \param NewType The new type to replace the references to. It is added to
+///        the model implicitly, no need to do so before calling this function.
+/// \param Model The tuple tree where replacement should take place in.
+///
+/// \return The new path to the added type.
+static model::TypePath
+fullTypeReplacement(const model::Type::Key &OldKey,
+                    UpcastablePointer<model::Type> &&NewType,
+                    TupleTree<model::Binary> &Model) {
+  // A temporary workaround to the problem of inability to replace constant
+  // references.
+  Model.initializeReferences();
```

Comment:
The model should not contain const references, I think.
`TupleTreeReference`s should contain `const` only out of model.
Isn't this the case?

Reply:
After speaking with @pfez, we want to introduce this in `verify` and he's going to implement that.
In the meantime, you can assume that.

Reply:
We discussed this extensively in the mattermost discussion the other day. But in the end the problem ended up being even more complex that it seems. It's almost like the const model is not usable in the general case.

Consider an example:
```cpp
model::TypedefType buildExample(const model::Binary &Model) {
  model::TypedefType IntPointer;
  IntPointer.UnderlyingType = { Model.getPrimitiveType(model::PrimitiveTypeKind::Unsigned, 4), {}};
  IntPointer.UnderlyingType = IntPointer.UnderlyingType.getPointerTo(Model.Architecture);
  return IntPointer;
}
```
Except the caller of the `buildExample` cannot then `record` this new type into the model, can they? Because the `UnderlyingType` contains a const root reference.

Reply:
`Model` should not be `const` there.
In general, if you're *constructing* a `model::Type` you need non-`const` model.
A `model::Type` off-model (i.e., not attached to a model) doesn't make sense. A `QualifiedType` does makes sense off-model, but a `model::Type` doesn't.

Maybe we can come up with a safer interface preventing this, but this is how things are right now. The bug there is creating a `TypedefType` with a `TupleTreeReference` containing a `const` `Root`.
`TypedefType.verify()` on that thing should fail (even if it doesn't as of now).

Reply:
This kind of makes me sad, because this means we cannot safely make something outside the model to evaluate if it even CAN be built to only them merge it in. But I see your point.

Removing a bunch of `const` all over the place makes me sad, but since it's not adjacent to any public interfaces, it should probably be fine (a typical user/fork would probably never touch this code after all).

Reply:
Eh, I understand. But I don't see an easy way out of this.

Reply:
Me too, me too.

Reply:
This is gone.